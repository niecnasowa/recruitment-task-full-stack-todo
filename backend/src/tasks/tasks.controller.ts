import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import { Task } from './task.model';
import { TasksService } from './tasks.service';
import { CreateTaskDto, UpdateTaskDto } from './tasks.dto';

@Controller('tasks')
export class TasksController {
  constructor(private tasksService: TasksService) {}

  @Get()
  findAll(): Task[] {
    return this.tasksService.findAll();
  }

  @Post()
  create(@Body() createTaskDto: CreateTaskDto): Task {
    return this.tasksService.create(createTaskDto.content);
  }

  @Delete(':id')
  delete(@Param('id') id: number): void {
    this.tasksService.delete(id);
  }

  @Patch(':id')
  update(@Param('id') id: number, @Body() updateTaskDto: UpdateTaskDto): Task {
    return this.tasksService.update(id, updateTaskDto.done);
  }
}
