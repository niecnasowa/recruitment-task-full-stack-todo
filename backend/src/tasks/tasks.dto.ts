import { IsBoolean, IsString, Length } from 'class-validator';

export class CreateTaskDto {
  @IsString()
  @Length(1, 9)
  content: string;
}

export class UpdateTaskDto {
  @IsBoolean()
  done: boolean;
}
