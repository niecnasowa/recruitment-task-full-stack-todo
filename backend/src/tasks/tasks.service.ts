import { Injectable, NotFoundException } from '@nestjs/common';
import { Task } from './task.model';

@Injectable()
export class TasksService {
  private tasks: Task[] = [
    { id: 1, content: 'content 1', done: false },
    { id: 2, content: 'content 2', done: true },
  ];

  findAll(): Task[] {
    return this.tasks;
  }

  create(content: string): Task {
    const task: Task = {
      id: Date.now(),
      content,
      done: false,
    };
    this.tasks.push(task);
    return task;
  }

  delete(id: number): void {
    const task = this.tasks.find((task) => task.id == id);
    if (!task) {
      throw new NotFoundException(`Task with ID ${id} not found`);
    }
    this.tasks = this.tasks.filter((task) => task.id != id);
  }

  update(id: number, done: boolean): Task {
    const task = this.tasks.find((task) => task.id == id);
    if (!task) {
      throw new NotFoundException(`Task with ID ${id} not found`);
    }
    if (task) task.done = done;
    return task;
  }
}
