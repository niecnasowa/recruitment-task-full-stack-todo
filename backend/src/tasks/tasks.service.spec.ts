import { Test, TestingModule } from '@nestjs/testing';
import { TasksService } from './tasks.service';

describe('TasksService', () => {
  let service: TasksService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [TasksService],
    }).compile();

    service = module.get<TasksService>(TasksService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('findAll', () => {
    it('should return an array of tasks', async () => {
      // Service has default value, so should be correct according to the default value
      expect(await service.findAll()).toMatchSnapshot();
    });
  });

  describe('create', () => {
    it('should create and return a task, should return this task when return all', async () => {
      const result = await service.create('Test task');

      expect(result).toMatchObject({
        content: 'Test task',
        done: false,
      });

      expect(await service.findAll()).toContainEqual(result);
    });
  });

  describe('delete', () => {
    it('should delete a task', async () => {
      expect((await service.findAll()).length).toBe(2);
      await service.delete(1);
      expect((await service.findAll()).length).toBe(1);
    });

    it('should throw an error if task is not found', async () => {
      expect((await service.findAll()).length).toBe(2);
      expect(() => service.delete(999)).toThrow('Task with ID 999 not found');
      expect((await service.findAll()).length).toBe(2);
    });
  });

  describe('update', () => {
    it('should update a task', async () => {
      const taskId = 1;

      expect(
        (await service.findAll()).find(({ id }) => id === taskId).done,
      ).toBe(false);

      expect(await service.update(taskId, true)).toMatchObject({ done: true });

      expect(
        (await service.findAll()).find(({ id }) => id === taskId).done,
      ).toBe(true);
    });

    it('should throw an error if task is not found', async () => {
      const taskId = 999;
      const allTasksBeforeAction = await service.findAll();

      expect(() => service.update(taskId, true)).toThrow(
        'Task with ID 999 not found',
      );
      expect(await service.findAll()).toEqual(allTasksBeforeAction);
    });
  });
});
