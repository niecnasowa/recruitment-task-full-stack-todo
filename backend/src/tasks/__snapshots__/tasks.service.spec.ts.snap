// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksService create should create and return a task 1`] = `
{
  "content": "Test task",
  "done": false,
  "id": 1694749621601,
}
`;

exports[`TasksService findAll should return an array of tasks 1`] = `
[
  {
    "content": "content 1",
    "done": false,
    "id": 1,
  },
  {
    "content": "content 2",
    "done": true,
    "id": 2,
  },
]
`;
